In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

In [3]:
results               = pd.read_csv('../data/raw/results.csv')
races                 = pd.read_csv('../data/raw/races.csv')
drivers               = pd.read_csv('../data/raw/drivers.csv')
constructors          = pd.read_csv('../data/raw/constructors.csv')
qualifying            = pd.read_csv('../data/raw/qualifying.csv')
status                = pd.read_csv('../data/raw/status.csv')
circuits              = pd.read_csv('../data/raw/circuits.csv')
lap_times             = pd.read_csv('../data/raw/lap_times.csv')
pit_stops             = pd.read_csv('../data/raw/pit_stops.csv')
driver_standings      = pd.read_csv('../data/raw/driver_standings.csv')
constructor_standings = pd.read_csv('../data/raw/constructor_standings.csv')
constructor_results   = pd.read_csv('../data/raw/constructor_results.csv')
seasons               = pd.read_csv('../data/raw/seasons.csv')
sprint_results        = pd.read_csv('../data/raw/sprint_results.csv')

In [4]:
print(f"Results: {results.shape}")
print(f"Races: {races.shape}")
print(f"Drivers: {drivers.shape}")
print(f"Constructors: {constructors.shape}")
print(f"Qualifying: {qualifying.shape}")
print(f"Status: {status.shape}")
print(f"Circuits: {circuits.shape}")
print(f"Lap Times: {lap_times.shape}")
print(f"Pit Stops: {pit_stops.shape}")
print(f"Driver Standings: {driver_standings.shape}")
print(f"Constructor Standings: {constructor_standings.shape}")
print(f"Constructor Results: {constructor_results.shape}")
print(f"Seasons: {seasons.shape}")
print(f"Sprint Results: {sprint_results.shape}")


Results: (26759, 18)
Races: (1125, 18)
Drivers: (861, 9)
Constructors: (212, 5)
Qualifying: (10494, 9)
Status: (139, 2)
Circuits: (77, 9)
Lap Times: (589081, 6)
Pit Stops: (11371, 7)
Driver Standings: (34863, 7)
Constructor Standings: (13391, 7)
Constructor Results: (12625, 5)
Seasons: (75, 2)
Sprint Results: (360, 16)


In [5]:
results.head()

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1,1,1,1,10.0,58,1:34:50.616,5690616,39,2,1:27.452,218.300,1
1,2,18,2,2,3,5,2,2,2,8.0,58,+5.478,5696094,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7,3,3,3,6.0,58,+8.163,5698779,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11,4,4,4,5.0,58,+17.181,5707797,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3,5,5,5,4.0,58,+18.014,5708630,43,1,1:27.418,218.385,1


In [6]:
races.head()

,raceId,year,round,circuitId,name,date,time,url,fp1_date,fp1_time,fp2_date,fp2_time,fp3_date,fp3_time,quali_date,quali_time,sprint_date,sprint_time
0,1,2009,1,1,Australian Grand Prix,2009-03-29,06:00:00,http://en.wikipedia.org/wiki/2009_Australian_G...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
1,2,2009,2,2,Malaysian Grand Prix,2009-04-05,09:00:00,http://en.wikipedia.org/wiki/2009_Malaysian_Gr...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
2,3,2009,3,17,Chinese Grand Prix,2009-04-19,07:00:00,http://en.wikipedia.org/wiki/2009_Chinese_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
3,4,2009,4,3,Bahrain Grand Prix,2009-04-26,12:00:00,http://en.wikipedia.org/wiki/2009_Bahrain_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
4,5,2009,5,4,Spanish Grand Prix,2009-05-10,12:00:00,http://en.wikipedia.org/wiki/2009_Spanish_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N


In [7]:
drivers.head()

,driverId,driverRef,number,code,forename,surname,dob,nationality,url
0,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen


In [8]:
constructors.head()

,constructorId,constructorRef,name,nationality,url
0,1,mclaren,McLaren,British,http://en.wikipedia.org/wiki/McLaren
1,2,bmw_sauber,BMW Sauber,German,http://en.wikipedia.org/wiki/BMW_Sauber
2,3,williams,Williams,British,http://en.wikipedia.org/wiki/Williams_Grand_Pr...
3,4,renault,Renault,French,http://en.wikipedia.org/wiki/Renault_in_Formul...
4,5,toro_rosso,Toro Rosso,Italian,http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso


In [9]:
qualifying.head()

,qualifyId,raceId,driverId,constructorId,number,position,q1,q2,q3
0,1,18,1,1,22,1,1:26.572,1:25.187,1:26.714
1,2,18,9,2,4,2,1:26.103,1:25.315,1:26.869
2,3,18,5,1,23,3,1:25.664,1:25.452,1:27.079
3,4,18,13,6,2,4,1:25.994,1:25.691,1:27.178
4,5,18,2,2,3,5,1:25.960,1:25.518,1:27.236


In [10]:
driver_standings.head()

,driverStandingsId,raceId,driverId,points,position,positionText,wins
0,1,18,1,10.0,1,1,1
1,2,18,2,8.0,2,2,0
2,3,18,3,6.0,3,3,0
3,4,18,4,5.0,4,4,0
4,5,18,5,4.0,5,5,0


In [11]:
constructor_standings.head()

,constructorStandingsId,raceId,constructorId,points,position,positionText,wins
0,1,18,1,14.0,1,1,1
1,2,18,2,8.0,3,3,0
2,3,18,3,9.0,2,2,0
3,4,18,4,5.0,4,4,0
4,5,18,5,2.0,5,5,0


In [12]:
pit_stops.head()

,raceId,driverId,stop,lap,time,duration,milliseconds
0,841,153,1,1,17:05:23,26.898,26898
1,841,30,1,1,17:05:52,25.021,25021
2,841,17,1,11,17:20:48,23.426,23426
3,841,4,1,12,17:22:34,23.251,23251
4,841,13,1,13,17:24:10,23.842,23842


In [13]:
status.head()

,statusId,status
0,1,Finished
1,2,Disqualified
2,3,Accident
3,4,Collision
4,5,Engine


# 🔗 Step 2: Data Merging

Joining all CSV files into one main dataframe.

## Why are we merging?
- `results.csv` is the **base** of our project
- It contains IDs like `raceId`, `driverId`, `constructorId` — not actual names
- We need to join other files to get **actual information**

## Merge Order:
1. results + races → add year, circuit, round
2. + drivers → add driver name, nationality
3. + constructors → add team name
4. + status → add Finished/DNF info
5. + qualifying → add qualifying position
6. + driver_standings → driver championship position before race
7. + constructor_standings → team championship position before race
8. + pit_stops → how many times driver pitted

In [14]:
# Step 1: results + races
# Adding year, round and circuit info to results
df = results.merge(
    races[['raceId', 'year', 'round', 'circuitId', 'name']], 
    on='raceId', how='left'
)
print(f"Step 1 done: {df.shape}")

# Step 2: Add driver information
# Adding driver name and nationality
df = df.merge(
    drivers[['driverId', 'code', 'forename', 'surname', 'nationality']], 
    on='driverId', how='left'
)
print(f"Step 2 done: {df.shape}")

# Step 3: Add team name
# Replacing constructorId with actual team name
df = df.merge(
    constructors[['constructorId', 'name']], 
    on='constructorId', how='left',
    suffixes=('_circuit', '_team')
)
print(f"Step 3 done: {df.shape}")

# Step 4: Add race status
# Adding Finished / DNF / Accident etc.
df = df.merge(
    status[['statusId', 'status']], 
    on='statusId', how='left'
)
print(f"Step 4 done: {df.shape}")

# Step 5: Add qualifying position
# Where was the driver in qualifying before the race
quali_clean = qualifying[['raceId', 'driverId', 'position']].rename(
    columns={'position': 'quali_position'}
)
df = df.merge(quali_clean, on=['raceId', 'driverId'], how='left')
print(f"Step 5 done: {df.shape}")

# Step 6: Add driver championship standings
# Driver's points and position in championship before this race
ds_clean = driver_standings[['raceId', 'driverId', 'points', 'position']].rename(
    columns={'points': 'driver_pts_before', 
             'position': 'driver_standing_before'}
)
df = df.merge(ds_clean, on=['raceId', 'driverId'], how='left')
print(f"Step 6 done: {df.shape}")

# Step 7: Add constructor championship standings
# Team's points and position in championship before this race
cs_clean = constructor_standings[['raceId', 'constructorId', 'points', 'position']].rename(
    columns={'points': 'team_pts_before', 
             'position': 'team_standing_before'}
)
df = df.merge(cs_clean, on=['raceId', 'constructorId'], how='left')
print(f"Step 7 done: {df.shape}")

# Step 8: Add pit stop count
# How many times each driver pitted in the race
pit_count = pit_stops.groupby(
    ['raceId', 'driverId']
).size().reset_index(name='pit_stop_count')

df = df.merge(pit_count, on=['raceId', 'driverId'], how='left')
print(f"Step 8 done: {df.shape}")

print("\n Main dataframe ready!")
print(f"Total rows    : {df.shape[0]}")
print(f"Total columns : {df.shape[1]}")
print(f"\nAll Columns:\n{df.columns.tolist()}")

Step 1 done: (26759, 22)
Step 2 done: (26759, 26)
Step 3 done: (26759, 27)
Step 4 done: (26759, 28)
Step 5 done: (26759, 29)
Step 6 done: (26759, 31)
Step 7 done: (26759, 33)
Step 8 done: (26759, 34)

 Main dataframe ready!
Total rows    : 26759
Total columns : 34

All Columns:
['resultId', 'raceId', 'driverId', 'constructorId', 'number', 'grid', 'position', 'positionText', 'positionOrder', 'points', 'laps', 'time', 'milliseconds', 'fastestLap', 'rank', 'fastestLapTime', 'fastestLapSpeed', 'statusId', 'year', 'round', 'circuitId', 'name_circuit', 'code', 'forename', 'surname', 'nationality', 'name_team', 'status', 'quali_position', 'driver_pts_before', 'driver_standing_before', 'team_pts_before', 'team_standing_before', 'pit_stop_count']


# Step 3: First Look at the Data

Let's see what our merged dataframe looks like.
- Check shape, columns, data types
- See first few rows
- Understand what each column contains

In [15]:
df.head()

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,...,surname,nationality,name_team,status,quali_position,driver_pts_before,driver_standing_before,team_pts_before,team_standing_before,pit_stop_count
0,1,18,1,1,22,1,1,1,1,10.0,...,Hamilton,British,McLaren,Finished,1.0,10.0,1.0,14.0,1.0,NaN
1,2,18,2,2,3,5,2,2,2,8.0,...,Heidfeld,German,BMW Sauber,Finished,5.0,8.0,2.0,8.0,3.0,NaN
2,3,18,3,3,7,7,3,3,3,6.0,...,Rosberg,German,Williams,Finished,7.0,6.0,3.0,9.0,2.0,NaN
3,4,18,4,4,5,11,4,4,4,5.0,...,Alonso,Spanish,Renault,Finished,12.0,5.0,4.0,5.0,4.0,NaN
4,5,18,5,1,23,3,5,5,5,4.0,...,Kovalainen,Finnish,McLaren,Finished,3.0,4.0,5.0,14.0,1.0,NaN


In [16]:
print("\n == Data Types ==")
print(df.dtypes)


 == Data Types ==
resultId                    int64
raceId                      int64
driverId                    int64
constructorId               int64
number                        str
grid                        int64
position                      str
positionText                  str
positionOrder               int64
points                    float64
laps                        int64
time                          str
milliseconds                  str
fastestLap                    str
rank                          str
fastestLapTime                str
fastestLapSpeed               str
statusId                    int64
year                        int64
round                       int64
circuitId                   int64
name_circuit                  str
code                          str
forename                      str
surname                       str
nationality                   str
name_team                     str
status                        str
quali_position            flo

In [17]:
print("\n=== Basic Info ===")
print(df.info())


=== Basic Info ===
<class 'pandas.DataFrame'>
RangeIndex: 26759 entries, 0 to 26758
Data columns (total 34 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   resultId                26759 non-null  int64  
 1   raceId                  26759 non-null  int64  
 2   driverId                26759 non-null  int64  
 3   constructorId           26759 non-null  int64  
 4   number                  26759 non-null  str    
 5   grid                    26759 non-null  int64  
 6   position                26759 non-null  str    
 7   positionText            26759 non-null  str    
 8   positionOrder           26759 non-null  int64  
 9   points                  26759 non-null  float64
 10  laps                    26759 non-null  int64  
 11  time                    26759 non-null  str    
 12  milliseconds            26759 non-null  str    
 13  fastestLap              26759 non-null  str    
 14  rank                    26759

In [18]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
resultId,26759.0,13380.977391,7726.134642,1.0,6690.5,13380.0,20069.5,26764.0
raceId,26759.0,551.687283,313.265036,1.0,300.0,531.0,811.0,1144.0
driverId,26759.0,278.673530,282.703039,1.0,57.0,172.0,399.5,862.0
constructorId,26759.0,50.180537,61.551498,1.0,6.0,25.0,63.0,215.0
grid,26759.0,11.134796,7.202860,0.0,5.0,11.0,17.0,34.0
positionOrder,26759.0,12.794051,7.665951,1.0,6.0,12.0,18.0,39.0
points,26759.0,1.987632,4.351209,0.0,0.0,0.0,2.0,50.0
laps,26759.0,46.301768,29.496557,0.0,23.0,53.0,66.0,200.0
statusId,26759.0,17.224971,26.026104,1.0,1.0,10.0,14.0,141.0
year,26759.0,1991.394372,19.952885,1950.0,1977.0,1991.0,2009.0,2024.0


# Step 4: Checking Missing Values (Null Analysis)

Before building our model, we need to understand:
- Which columns have missing values?
- How many values are missing?
- Why are they missing?
- What should we do about them?

In [19]:
# Total nulls in each column
null_counts = df.isnull().sum()
null_percent = (df.isnull().sum() / len(df) * 100).round(2)

null_df = pd.DataFrame({
    'Missing Count': null_counts,
    'Missing %': null_percent
}).sort_values('Missing %', ascending=False)

# Only show columns that have nulls
null_df = null_df[null_df['Missing Count'] > 0]

print("=== Columns with Missing Values ===")
print(null_df)

=== Columns with Missing Values ===
                        Missing Count  Missing %
pit_stop_count                  21184      79.17
quali_position                  16265      60.78
team_pts_before                  1867       6.98
team_standing_before             1867       6.98
driver_pts_before                 469       1.75
driver_standing_before            469       1.75


#  Step 5: Visualizing Missing Values Distribution

Before filling missing values, let's understand:
- Where exactly are the missing values?
- Are they missing because of year (old data)?
- What pattern do they follow?

This will help us decide the best strategy to handle them.

In [20]:
# ── Graph 1: Overall Missing % Bar Chart ─────────────────
fig1 = go.Figure(go.Bar(
    x=null_df.index,
    y=null_df['Missing %'],
    marker_color=['#e10600','#ff6b6b','#ffa07a',
                  '#ffd700','#90EE90','#87CEEB'],
    text=[f'{v:.1f}%' for v in null_df['Missing %']],
    textposition='outside'
))
fig1.update_layout(
    title='Overall Missing Values % per Column',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#e10600', size=20),
    xaxis_title='Column',
    yaxis_title='Missing %',
    yaxis=dict(gridcolor='#333333')
)
fig1.show()

# ── Graph 2: pit_stop_count missing by year ───────────────
pit_by_year = df.groupby('year')['pit_stop_count'].apply(
    lambda x: x.isnull().sum()
).reset_index()
pit_by_year.columns = ['year', 'missing_count']

fig2 = go.Figure(go.Bar(
    x=pit_by_year['year'],
    y=pit_by_year['missing_count'],
    marker_color='#e10600'
))
fig2.update_layout(
    title='pit_stop_count — Missing by Year<br><sup>➡ Data only available from 2011</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#e10600', size=18),
    xaxis_title='Year',
    yaxis_title='Missing Count',
    yaxis=dict(gridcolor='#333333')
)
fig2.show()

# ── Graph 3: quali_position missing by year ───────────────
quali_by_year = df.groupby('year')['quali_position'].apply(
    lambda x: x.isnull().sum()
).reset_index()
quali_by_year.columns = ['year', 'missing_count']

fig3 = go.Figure(go.Bar(
    x=quali_by_year['year'],
    y=quali_by_year['missing_count'],
    marker_color='#ffd700'
))
fig3.update_layout(
    title='quali_position — Missing by Year<br><sup>➡ Old seasons had no qualifying data</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#e10600', size=18),
    xaxis_title='Year',
    yaxis_title='Missing Count',
    yaxis=dict(gridcolor='#333333')
)
fig3.show()

# ── Graph 4: driver_pts_before missing by year ────────────
pts_by_year = df.groupby('year')['driver_pts_before'].apply(
    lambda x: x.isnull().sum()
).reset_index()
pts_by_year.columns = ['year', 'missing_count']

fig4 = go.Figure(go.Bar(
    x=pts_by_year['year'],
    y=pts_by_year['missing_count'],
    marker_color='#00ff7f'
))
fig4.update_layout(
    title='driver_pts_before — Missing by Year<br><sup>➡ Only first race of each season</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#e10600', size=18),
    xaxis_title='Year',
    yaxis_title='Missing Count',
    yaxis=dict(gridcolor='#333333')
)
fig4.show()

# ── Graph 5: Heatmap — missing pattern ───────────────────
missing_by_year = df.groupby('year')[null_df.index.tolist()].apply(
    lambda x: x.isnull().sum()
)

fig5 = go.Figure(go.Heatmap(
    z=missing_by_year.values,
    x=missing_by_year.columns,
    y=missing_by_year.index,
    colorscale=[
        [0, '#1a1a2e'],
        [0.5, '#ffd700'],
        [1, '#e10600']
    ],
    text=missing_by_year.values,
    texttemplate='%{text}',
))
fig5.update_layout(
    title='Missing Values Heatmap — Year vs Column',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#e10600', size=20),
    xaxis_title='Column',
    yaxis_title='Year'
)
fig5.show()

#  Step 6: Handling Missing Values

## Strategy for each column:

| Column | Missing % | Why? | Solution |
|--------|-----------|------|----------|
| `pit_stop_count` | 79% | Data only from 2011 | Fill with 0 for old races, or filter year >= 2011 |
| `quali_position` | 61% | Old seasons had no qualifying | Fill with `grid` position |
| `team_pts_before` | 7% | First race of season | Fill with 0 |
| `driver_pts_before` | 2% | First race of season | Fill with 0 |
| `team_standing_before` | 7% | First race of season | Fill with 0 |
| `driver_standing_before` | 2% | First race of season | Fill with 0 |

In [21]:
df = df[df['year']>= 2000].copy()
print(f"After filtering year : {df.shape}")

# Strategy 2 — quali_position NaN →  fill with grid position
df['quali_position'] = df['quali_position'].fillna(df['grid'])
print(f"quali_position nulls : {df['quali_position'].isnull().sum()}")

# Strategy 3 — points/standings NaN → fill with 0
df['driver_pts_before'] = df['driver_pts_before'].fillna(0)
df['driver_standing_before'] = df['driver_standing_before'].fillna(0)
df['team_pts_before']        = df['team_pts_before'].fillna(0)
df['team_standing_before']   = df['team_standing_before'].fillna(0)
print("points/standings nulls filled with 0")

After filtering year : (10079, 34)
quali_position nulls : 0
points/standings nulls filled with 0


# Step 7: Exploratory Data Analysis (EDA)

## 7.1 Target Variable — positionOrder
Our target is `positionOrder` — the finishing position (1 to 20).
Let's understand its distribution first.

In [22]:
fig  = go.Figure()

fig.add_trace(go.Histogram(
    x=df['positionOrder'],
    nbinsx=20,
    marker_color='#e10600',
    opacity=0.85,
    name='Finish Position'
))

fig.update_layout(
    title = 'Distribution of Finish Positions',
    paper_bgcolor = "#1a1a2e",
    plot_bgcolor = "#1a1a2e",
    font = dict(color = 'white'),
    title_font = dict(color = '#e10600', size = 20),
    xaxis_title = 'Finish Position',
    yaxis_title = 'Count',
    xaxis = dict(gridcolor='#333333'),
    yaxis = dict(gridcolor='#333333'),
    bargap=0.2
)
fig.show()

print(df['positionOrder'].describe())

count    10079.000000
mean        11.069551
std          6.154428
min          1.000000
25%          6.000000
50%         11.000000
75%         16.000000
max         24.000000
Name: positionOrder, dtype: float64


## 6.2 Grid Position Distribution

Grid position = where the driver starts the race.
- P1 = Pole position (fastest in qualifying)
- P20 = Last on grid
This is our most important feature!

In [23]:
fig = go.Figure(go.Bar(
    x=list(range(1, 21)),
    y=df[df['grid'].between(1,20)]['grid'].value_counts().sort_index().values,
    marker=dict(
        color=list(range(1, 21)),
        colorscale='Viridis',
        showscale=True,
        colorbar=dict(title='Grid Position')
    ),
    opacity=0.9
))

fig.update_layout(
    title='Distribution of Grid Positions',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Grid Position',
    yaxis_title='Count',
    xaxis=dict(gridcolor='#333333', dtick=1),
    yaxis=dict(gridcolor='#333333'),
    bargap=0.1
)
fig.show()

## 6.2 Grid Position Distribution

### What does this graph show?
Each bar = how many times drivers started from that grid position
across all races from 1950 to 2023.

### Key Observations:

**1. P1 to P18 — Almost Equal Distribution**
All positions from P1 to P18 have roughly equal count (~470 each).
This makes sense — in every race, each grid slot is filled once.

**2. P19 and P20 — Slightly Lower**
These positions have fewer counts (~420 and ~250) because:
- In older seasons (1950s-1990s) fewer than 20 cars participated
- Grid size was not always 20 cars

**3. Positions beyond P20 exist**
A small bar appears after P20 — these are from old seasons
where grids had 22, 24, or even 26 cars.

**4. Uniform distribution confirms data quality**
The near-equal distribution across P1-P18 tells us our
data is consistent and well-structured.

### Conclusion:
> Grid position data is evenly distributed and reliable.
> No single starting position is over or under-represented
> in our dataset — this is ideal for ML training.

## 6.3 Grid Position vs Finish Position

This is the most important relationship in our project!
- Does starting position determine finishing position?
- Where do upsets happen?
- Strong correlation = grid is a powerful feature

In [24]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=df['grid'],
    y=df['positionOrder'],
    mode='markers',
    marker=dict(
        color=df['year'],
        colorscale='Viridis',
        showscale=True,
        colorbar=dict(title='Year'),
        size=3,
        opacity=0.5
    ),
    text=df['code'],
    hovertemplate='<b>%{text}</b><br>Grid: %{x}<br>Finish: %{y}<extra></extra>'
))

# Diagonal line — perfect prediction line
fig.add_trace(go.Scatter(
    x=list(range(1, 21)),
    y=list(range(1, 21)),
    mode='lines',
    line=dict(color='#e10600', width=2, dash='dash'),
    name='Perfect prediction'
))

fig.update_layout(
    title='Grid Position vs Finish Position<br><sup>Points below red line = gained positions, above = lost positions</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Grid Position',
    yaxis_title='Finish Position',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig.show()

# Correlation check
corr = df['grid'].corr(df['positionOrder'])
print(f"Correlation between Grid and Finish: {corr:.3f}")
print("0 = no relation, 1 = perfect relation")

Correlation between Grid and Finish: 0.556
0 = no relation, 1 = perfect relation


### Key Observations:

**1. Moderate-Strong Correlation (0.556)**
Grid position has a meaningful impact on finish position
but it is not the only factor. Other variables like
team strength, driver skill, and strategy also matter.

**2. Pole Position Advantage is Real**
P1 starters finish P1 or P2 more often than any other
grid slot — top-left corner of scatter is heavily populated.

**3. Mid-Grid is Most Unpredictable**
Drivers starting P8 to P15 show the most scatter —
safety cars, pit strategy, and overtaking create chaos here.

**4. Back of Grid Rarely Reaches Podium**
Starting P15+ and finishing P1-P3 is extremely rare —
only happens in wet races or major incidents ahead.

**5. Modern F1 More Predictable**
Recent seasons (yellow dots, 2018-2023) cluster tighter
around the diagonal — aerodynamic complexity makes
overtaking harder, so grid position matters more now.

### Conclusion:
> With a correlation of **0.556**, grid position is our
> **single strongest predictor** of race finish.
> However, 0.556 also tells us the model needs
> additional features to improve predictions —
> qualifying position, team strength, and driver
> form will all add predictive power.

## 6.4 Qualifying Position vs Finish Position

Qualifying position = fastest lap time in qualifying session.
This tells us how fast the car + driver combination is
before the race even starts.

**Key Question:** Is qualifying more predictive than grid?
Let's find out!

In [25]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=df['quali_position'],
    y=df['positionOrder'],
    mode='markers',
    marker=dict(
        color=df['year'],
        colorscale='Plasma',
        showscale=True,
        colorbar=dict(title='Year'),
        size=3,
        opacity=0.5
    ),
    text=df['code'],
    hovertemplate='<b>%{text}</b><br>Quali: %{x}<br>Finish: %{y}<extra></extra>'
))

# Perfect prediction line
fig.add_trace(go.Scatter(
    x=list(range(1, 21)),
    y=list(range(1, 21)),
    mode='lines',
    line=dict(color='#00D2BE', width=2, dash='dash'),
    name='Perfect prediction'
))

fig.update_layout(
    title='Qualifying Position vs Finish Position<br><sup>How much does qualifying pace predict race result?</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Qualifying Position',
    yaxis_title='Finish Position',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig.show()

# Correlation
corr_quali = df['quali_position'].corr(df['positionOrder'])
corr_grid  = df['grid'].corr(df['positionOrder'])

print(f"Qualifying vs Finish correlation : {corr_quali:.3f}")
print(f"Grid vs Finish correlation       : {corr_grid:.3f}")
print(f"\nWhich is stronger? {'Qualifying' if corr_quali > corr_grid else 'Grid'}")

Qualifying vs Finish correlation : 0.588
Grid vs Finish correlation       : 0.556

Which is stronger? Qualifying



### Key Observations from Graph:

**1. Qualifying is Slightly More Predictive than Grid (0.588 > 0.556)**
Qualifying position has a marginally stronger relationship
with finish position than grid position.
This makes sense because qualifying reflects pure car +
driver pace, while grid can be affected by penalties.

**2. Strong clustering near the diagonal**
Most dots follow the teal dashed line closely —
especially in recent years (yellow dots, 2018-2023).
This confirms qualifying pace directly translates to race pace.

**3. Q1 starters dominate race wins**
Bottom-left corner (Quali P1, Finish P1-3) is heavily
populated — pole sitters win most often.

**4. Large spread in mid-field (Q8-Q15)**
These qualifying positions show the most scatter —
strategy, safety cars, and overtaking create chaos here.

**5. Modern F1 more predictable (yellow dots)**
Recent seasons cluster tighter around diagonal —
aerodynamic complexity makes it harder to overtake,
so qualifying position matters more than ever.

**6. Old seasons more chaotic (blue/purple dots)**
Pre-2010 races show more scatter — older cars were
easier to overtake, so qualifying mattered less.

**7. Why are Grid and Quali so similar?**
In most races, qualifying position = grid position.
Difference occurs when:
- Driver takes grid penalty (engine change etc.)
- Driver starts from pit lane
- Weather changes between qualifying and race

---

### Conclusion:
> Qualifying position **(0.588)** is our **strongest single
> predictor** — slightly better than grid (0.556).
> Using **both features together** will give our model
> the best predictive power for race finish position.
> The 0.032 difference also tells us grid penalties
> are a meaningful signal worth capturing separately.

## 6.5 Team Performance Analysis

Which constructors have dominated F1 across different eras?
- Average finish position per team
- Team consistency over seasons
- How team strength relates to driver finish position

In [26]:
team_avg = df.groupby('name_team')['positionOrder'].mean().reset_index()
team_avg.columns = ['team', 'avg_finish']
team_avg = team_avg.sort_values('avg_finish').head(10)

fig1 = go.Figure(go.Bar(
    x=team_avg['avg_finish'],
    y=team_avg['team'],
    orientation='h',
    marker=dict(
        color=team_avg['avg_finish'],
        colorscale='RdYlGn_r',
        showscale=True,
        colorbar=dict(title='Avg Finish')
    ),
    text=[f'P{v:.1f}' for v in team_avg['avg_finish']],
    textposition='outside'
))

fig1.update_layout(
    title='Top 10 Teams — Average Finish Position (All Time)<br><sup>Lower = Better</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Average Finish Position',
    yaxis_title='Team',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig1.show()

# Team performance over years — top 5 teams
top5_teams = team_avg['team'].head(5).tolist()
team_year = df[df['name_team'].isin(top5_teams)].groupby(
    ['year', 'name_team']
)['positionOrder'].mean().reset_index()

fig2 = px.line(
    team_year,
    x='year',
    y='positionOrder',
    color='name_team',
    title='Top 5 Teams — Average Finish Position Over Years',
    markers=True
)
fig2.update_yaxes(autorange='reversed')
fig2.update_layout(
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Year',
    yaxis_title='Avg Finish Position',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
    legend_title='Team'
)
fig2.show()

# Print stats
print("=== Top 10 Teams — Average Finish Position ===")
print(team_avg.to_string(index=False))

=== Top 10 Teams — Average Finish Position ===
        team  avg_finish
       Brawn    4.882353
    Mercedes    5.909836
     Ferrari    6.479123
    Red Bull    7.347716
  BMW Sauber    8.921429
     McLaren    8.981211
     Renault   10.467626
Racing Point   10.631579
         BAR   10.916667
Aston Martin   11.155556


## 6.6 Driver Performance Analysis

In [28]:
driver_stats = df.groupby('code').agg(
    avg_finish = ('positionOrder', 'mean'),
    total_races = ('positionOrder', 'count'),
    best_finish = ('positionOrder', 'min'),
    wins = ('positionOrder', lambda x : (x == 1).sum())
).reset_index()

river_stats = driver_stats[driver_stats['total_races'] >= 50]
top10 = driver_stats.sort_values('avg_finish').head(10)

# Graph 1 — Top 10 drivers bar chart
fig1 = go.Figure(go.Bar(
    x=top10['avg_finish'],
    y=top10['code'],
    orientation='h',
    marker=dict(
        color=top10['avg_finish'],
        colorscale='RdYlGn_r',
        showscale=True,
        colorbar=dict(title='Avg Finish')
    ),
    text=[f'P{v:.1f} ({w} wins)' for v, w in 
          zip(top10['avg_finish'], top10['wins'])],
    textposition='outside'
))

fig1.update_layout(
    title='Top 10 Drivers — Average Finish Position (Min 50 races)<br><sup>Lower = Better</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Average Finish Position',
    yaxis_title='Driver',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig1.show()

# Graph 2 — Top 5 drivers over years
top5_drivers = top10['code'].head(5).tolist()
driver_year = df[df['code'].isin(top5_drivers)].groupby(
    ['year', 'code']
)['positionOrder'].mean().reset_index()

fig2 = px.line(
    driver_year,
    x='year',
    y='positionOrder',
    color='code',
    title='Top 5 Drivers — Performance Over Years',
    markers=True
)
fig2.update_yaxes(autorange='reversed')
fig2.update_layout(
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Year',
    yaxis_title='Avg Finish Position',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
    legend_title='Driver'
)
fig2.show()

print("=== Top 10 Drivers Stats ===")
print(top10[['code', 'avg_finish', 'total_races', 'wins']].to_string(index=False))

=== Top 10 Drivers Stats ===
code  avg_finish  total_races  wins
 HAM    5.019663          356   105
 VET    7.093333          300    53
 VER    7.456929          267    63
 LEC    7.557047          149     8
 NOR    7.851562          128     4
 PIA    7.891304           46     2
 ROS    8.252427          206    23
 RAI    8.491477          352    21
 ALO    8.492574          404    32
 MON    8.526316           95     7


### Key Points:

**1. Hamilton — GOAT of Consistency**
Best average finish (P5.0) across 356 races with 105 wins.
No other driver comes close in terms of consistency + volume.

**2. Verstappen — Most Wins per Race**
63 wins in just 267 races = highest win rate among top drivers.
Shows recent Red Bull dominance (2022-2023).

**3. Vettel vs Verstappen**
VET: 53 wins, 300 races — Red Bull 2010-2013 era
VER: 63 wins, 267 races — already surpassed Vettel in fewer races

**4. Young Guns Rising (LEC, NOR)**
LEC and NOR in top 10 with fewer races — 
their averages will improve as careers develop.

**5. Line Chart — Three Clear Eras**
- 2006-2013 → HAM + VET dominant
- 2014-2021 → HAM dominant (Mercedes era)
- 2022-2024 → VER takes over (Red Bull era)

### Conclusion:
> Driver identity is a **strong feature** for our model.
> HAM, VET, VER consistently outperform their grid slots.

## 6.7 Pit Stop Analysis

In [29]:
df_pit = df[df['year'] >= 2011].copy()

# Graph 1 — Pit stop count distribution
pit_dist = df_pit['pit_stop_count'].value_counts().sort_index().reset_index()
pit_dist.columns = ['pit_stops', 'count']

fig1 = go.Figure(go.Bar(
    x=pit_dist['pit_stops'],
    y=pit_dist['count'],
    marker=dict(
        color=pit_dist['pit_stops'],
        colorscale='Viridis',
        showscale=True,
        colorbar=dict(title='Pit Stops')
    ),
    text=pit_dist['count'],
    textposition='outside'
))

fig1.update_layout(
    title='Pit Stop Count Distribution (2011-2023)',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Number of Pit Stops',
    yaxis_title='Count',
    xaxis=dict(gridcolor='#333333', dtick=1),
    yaxis=dict(gridcolor='#333333'),
    bargap=0.2
)
fig1.show()

# Graph 2 — Pit stops vs Finish position
pit_finish = df_pit.groupby('pit_stop_count')['positionOrder'].mean().reset_index()

fig2 = go.Figure(go.Scatter(
    x=pit_finish['pit_stop_count'],
    y=pit_finish['positionOrder'],
    mode='lines+markers',
    marker=dict(
        color=pit_finish['positionOrder'],
        colorscale='RdYlGn_r',
        size=12,
        showscale=True,
        colorbar=dict(title='Avg Finish')
    ),
    line=dict(color='#00D2BE', width=2),
    text=[f'P{v:.1f}' for v in pit_finish['positionOrder']],
    textposition='top center'
))

fig2.update_layout(
    title='Pit Stops vs Average Finish Position (2011-2023)<br><sup>Does more pit stops = better or worse finish?</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Number of Pit Stops',
    yaxis_title='Average Finish Position',
    xaxis=dict(gridcolor='#333333', dtick=1),
    yaxis=dict(gridcolor='#333333'),
)
fig2.show()

# Correlation
corr_pit = df_pit['pit_stop_count'].corr(df_pit['positionOrder'])
print(f"Pit stops vs Finish correlation: {corr_pit:.3f}")
print("\n=== Avg Finish by Pit Stops ===")
print(pit_finish.to_string(index=False))

Pit stops vs Finish correlation: 0.039

=== Avg Finish by Pit Stops ===
 pit_stop_count  positionOrder
            1.0      10.240022
            2.0      10.162697
            3.0      10.926099
            4.0      11.026059
            5.0      10.285714
            6.0      10.384615
            7.0      13.666667


### Correlation: 0.039 (Very Weak)

### Key Points:

**1. 2 Stops is Most Common Strategy (40%)**
Modern F1 tyre regulations favour 2-stop strategy —
mandatory compound change rule pushes teams here.

**2. Very Weak Correlation (0.039)**
Pit stop count barely affects finish position.
This tells us pit stops alone don't determine result —
it is the TIMING of stops that matters, not the count.

**3. More Stops = Slightly Worse Finish**
Drivers with 5+ stops average P13+ finish —
these are usually drivers who had problems,
punctures, or safety car strategy gambles.

**4. 1 vs 2 Stops — Almost Same Result (P10.2)**
One-stop and two-stop strategies produce identical
average finish — strategy choice depends on
tyre degradation and circuit characteristics.

**5. 7 Stops = P13.7 — Worst Average**
Extreme pit stop counts indicate race incidents —
punctures, damage, or desperate strategy changes.


## 6.8 Year Wise Trends

In [30]:
# Graph 1 — Grid vs Finish correlation by year
corr_by_year = df.groupby('year').apply(
    lambda x: x['grid'].corr(x['positionOrder'])
).reset_index()
corr_by_year.columns = ['year', 'correlation']

fig1 = go.Figure()

fig1.add_trace(go.Scatter(
    x=corr_by_year['year'],
    y=corr_by_year['correlation'],
    mode='lines+markers',
    marker=dict(
        color=corr_by_year['correlation'],
        colorscale='RdYlGn',
        size=8,
        showscale=True,
        colorbar=dict(title='Correlation')
    ),
    line=dict(color='#00D2BE', width=2),
    fill='tozeroy',
    fillcolor='rgba(0, 210, 190, 0.1)'
))

fig1.update_layout(
    title='Grid vs Finish Correlation by Year<br><sup>Higher = Grid position more predictive that year</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Year',
    yaxis_title='Correlation',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig1.show()

# Graph 2 — Average positions gained/lost per year
df['positions_gained'] = df['grid'] - df['positionOrder']
# Positive = gained positions, Negative = lost positions

yearly_gains = df.groupby('year')['positions_gained'].mean().reset_index()

fig2 = go.Figure(go.Bar(
    x=yearly_gains['year'],
    y=yearly_gains['positions_gained'],
    marker=dict(
        color=yearly_gains['positions_gained'],
        colorscale='RdYlGn',
        showscale=True,
        colorbar=dict(title='Avg Positions')
    ),
    text=[f'{v:.2f}' for v in yearly_gains['positions_gained']],
    textposition='outside'
))

fig2.update_layout(
    title='Average Positions Gained/Lost per Year<br><sup>Positive = drivers gained positions on average</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Year',
    yaxis_title='Avg Positions Gained',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333', zeroline=True,
               zerolinecolor='white', zerolinewidth=1),
    bargap=0.1
)
fig2.show()

# Graph 3 — Total races per year
races_per_year = df.groupby('year')['raceId'].nunique().reset_index()
races_per_year.columns = ['year', 'total_races']

fig3 = go.Figure(go.Scatter(
    x=races_per_year['year'],
    y=races_per_year['total_races'],
    mode='lines+markers',
    fill='tozeroy',
    fillcolor='rgba(225, 6, 0, 0.1)',
    line=dict(color='#e10600', width=2),
    marker=dict(color='#e10600', size=6)
))

fig3.update_layout(
    title='Total Races per Season (1950-2023)<br><sup>F1 calendar has grown significantly</sup>',
    paper_bgcolor='#1a1a2e',
    plot_bgcolor='#1a1a2e',
    font=dict(color='white'),
    title_font=dict(color='#00D2BE', size=20),
    xaxis_title='Year',
    yaxis_title='Number of Races',
    xaxis=dict(gridcolor='#333333'),
    yaxis=dict(gridcolor='#333333'),
)
fig3.show()

print("=== Recent Years Correlation ===")
print(corr_by_year.tail(10).to_string(index=False))

=== Recent Years Correlation ===
 year  correlation
 2015     0.612358
 2016     0.605333
 2017     0.553710
 2018     0.563812
 2019     0.567470
 2020     0.559700
 2021     0.578387
 2022     0.477903
 2023     0.496516
 2024     0.661817


## 6.8 Year Wise Trends

### Graph 1: Grid vs Finish Correlation by Year
- Overall correlation has **increased over time** (0.4 → 0.66)
- Modern F1 more predictable — harder to overtake
- **2022-23 dip (0.47-0.49)** — new regulations allowed more racing
- **2024 peak (0.661)** — most processional season ever

### Graph 2: Positions Gained/Lost
- All bars **negative** — drivers lose positions on average due to DNFs
- **2021 worst (-0.75)** — COVID calendar, unusual circuits
- Pre-2015 nearly zero — races were more processional

### Graph 3: Total Races per Season
- Calendar grew from **17 races (2000) → 24 races (2023)**
- **2020 dip** — COVID shortened season to 17 races
- Liberty Media era (2017+) aggressively expanded calendar

### Conclusion
> `year` is an important feature — F1 regulations and
> race predictability vary significantly across eras

In [33]:
import os

os.makedirs('../data/processed', exist_ok=True)

df.to_csv('../data/processed/merged.csv', index=False)
print(f" Saved! Shape: {df.shape}")
print("Location: data/processed/merged.csv")

 Saved! Shape: (10079, 35)
Location: data/processed/merged.csv
